### ISSUE TITLE
octet_counting framing requires data to be UTF-8 encoded

### ISSUE THREAD

**Issue by hanshuebner** [vectordotdev/vector#19971]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

Some of the sources support the `octet_counting` framing, which allows multiple events to be present in an input buffer without requiring a delimeter character like `\n`.  The decoder for this framing format [enforces](https://github.com/vectordotdev/vector/blob/9dd9907b356996d9bbb395fd4aea2a207c930914/lib/codecs/src/decoding/framing/octet_counting.rs#L173-L187) the data itself to be UTF-8 encoded, however.  This makes the framing decoder unsuitable for binary formats like protobuf.  This seems odd, as binary formats are what makes using a delimited format difficult.  What is the reason for the explicit check?

Removing the check from the code changes the behavior of the source as expected, i.e. decoding multiple protobuf message with `octet_counting` framing works fine.  Would it be an option to make checking the data for UTF-8 encoding be optional?

### Configuration

```text
[sources.protobuf_source]
type = "stdin"
framing.method = "octet_counting"
decoding.codec = "protobuf"
decoding.protobuf.desc_file = "Trace.desc"
decoding.protobuf.message_type = "kong.trace.Message"

[sinks.console_sink]
type = "console"
encoding.codec = "json"
inputs = ["protobuf_source"]
```

### Version

vector 0.37.0-custom-cae37e99d (aarch64-apple-darwin debug=full)

**Comment by jszwedko** [comment]
Hmm, yeah, that is a little odd. That framing is most often used with syslog messages, which are usually ASCII or UTF-8, which may why it was implemented like that, but I'm not seeing a reason it _needs_ to be. I'd be happy to see a PR for this if you want to take a shot at making the changes.

For binary frames, we actually usually recommend the `length_delimited` codec, which is fairly similar in spirit to octet encoding and shouldn't require UTF-8.

**Comment by hanshuebner** [comment]
Thank you for quickly getting back.  The `length_delimited` coded suits our requirements just fine, but it is not in the documentation (yet) so I thought `octet_counting` would be my only choice.

I can open a PR to fix `octet_counting`.  In my opinion, removing the UTF-8 check altogether would be the right thing, but I can also make it an option (`framing.octet_counting.require_utf8`?) that defaults to `true`.

**Comment by jszwedko** [comment]
`length_delimited` is actually a documented codec option. See https://vector.dev/docs/reference/configuration/sources/socket/#framing.method for example. Were you expecting to see it somewhere else though? I do wish we had a dedicated page on codecs to describe them in more detail.

> I can open a PR to fix `octet_counting`. In my opinion, removing the UTF-8 check altogether would be the right thing, but I can also make it an option (`framing.octet_counting.require_utf8`?) that defaults to `true`.

Hmm, that's a good question. II think it is _unlikely_ to break behavior for people since they are likely using this framing with the `syslog` codec which requires UTF-8:

https://github.com/vectordotdev/vector/blob/cae37e99d8dba79c943e9cdf6af862523141f71c/lib/codecs/src/decoding/format/syslog.rs#L276-L279

(though maybe it shouldn't since technically the `msg` part of a Syslog mesasge doesn't need to be UTF-8 per RFC5424)

However there is also no way to recreate the behavior if they really _do_ want to restrict the incoming data to UTF-8.
Given that, I think maybe we should go through a deprecation cycle to let people raise concerns before we remove the behavior entirely. This would look like:

- Add a `require_utf8` option that defaults to `true` for the next release, but document the behavior as deprecated
- Switch `require_utf8` to `false` in a following release, which should make the behavior more widespread so people can raise issues
- Remove the `require_utf8` option

If you want to take a shot, you could just do the first step and then we could handle the following steps in the future.

**Comment by hanshuebner** [comment]
> length_delimited is actually a documented codec option. See https://vector.dev/docs/reference/configuration/sources/socket/#framing.method for example. Were you expecting to see it somewhere else though?

You are right, I was no looking closely enough.  I missed `length_delimited` because it has no options and thus does not show in the sidebar, my bad.

I'll prepare a PR to make the proposed change.
